gitlabform/processors/project/files_processor.py

```diff
@@ -17,6 +17,72 @@
 
 
 class FilesProcessor(AbstractProcessor):
+    def pre_protection_check(
+    self,
+    project_and_group: str,
+    branch: str,
+    desired_protected_state: bool,
+    desired_push_access_level: str,
+    desired_merge_access_level: str,
+    ):
+        
+        try:
+            branch_info = self.gitlab.get_branch(project_and_group, branch)
+            current_protected_state = branch_info["protected"]
+
+            # Check if the branch needs changes to its protection settings
+            needs_unprotect = current_protected_state and (
+                branch_info["push_access_levels"][0]["access_level_description"]
+                != desired_push_access_level
+                or branch_info["merge_access_levels"][0]["access_level_description"]
+                != desired_merge_access_level
+            )
+
+            # If the branch is unprotected or should remain unprotected, skip protection changes
+            if not current_protected_state and not desired_protected_state:
+                print(
+                    f"Branch '{branch}' is already unprotected and should remain so. Skipping changes."
+                )
+                return False
+
+            if needs_unprotect:
+                print(f"Branch '{branch}' needs changes in protection settings.")
+                return True
+            else:
+                print(f"No changes needed on branch '{branch}'.")
+                return False
+
+        except NotFoundException:
+            print(f"Branch '{branch}' not found.")
+            return False
+    
+    def process_branch(self, project_and_group: str, branch: str, desired_protected_state: bool,
+                       desired_push_access_level: str, desired_merge_access_level: str):
+        """
+        This method checks if changes are needed on the branch before unprotecting or reprotecting it.
+        """
+        modify_protection = False
+        # Call pre_protection_check to determine if any protection changes are needed
+        if not self.pre_protection_check(project_and_group, branch, desired_protected_state,
+                                         desired_push_access_level, desired_merge_access_level):
+            debug(f"No work needed on branch '{branch}'. Skipping any protection changes.")
+        else:
+            debug(f"Branch '{branch}' requires updates. Unprotecting and applying changes.")
+            modify_protection = True
+        
+        if modify_protection:
+            # Unprotect the branch if necessary
+            self.branch_protector.unprotect_branch(project_and_group, branch)
+
+        try:
+            # Apply the changes to the branch
+            self._process_configuration(project_and_group, branch)
```

Comment:
_process_configuration is called by the abstract interface somewhere upstream so this implementation in process_branch won't end up being called